executor/cook/progress.py

```diff
@@ -0,0 +1,248 @@
+
+                    self.send_message(self.driver, progress_message, self.max_message_length)
+                    self.last_reported_time = time.time()
+                    self.last_progress_data = progress_data
+                else:
+                    logging.debug('Not sending progress data as enough time has not elapsed since last update')
+
+
+class ProgressWatcher(object):
+    """This class tails the output from the target file listening for progress messages.
+    The retrieve_progress_states generates all progress messages iteratively.
+    """
+
+    @staticmethod
+    def match_progress_update(progress_regex_str, input_string):
+        """Returns the progress tuple when the input string matches the provided regex.
+        
+        Parameters
+        ----------
+        progress_regex_str: string
+            The progress regex to match against, it must return two capture groups.
+        input_string: string
+            The input string.
+        
+        Returns
+        -------
+        the tuple (percent, message) if the string matches the provided regex, 
+                 else return None. 
+        """
+        matches = re.findall(progress_regex_str, input_string)
+        return matches[0] if len(matches) >= 1 else None
+
+    def __init__(self, config, completed_signal):
+        self.target_file = config.progress_output_name
+        self.progress_regex_string = config.progress_regex_string
+        self.progress = None
+        self.completed_signal = completed_signal
+        self.max_bytes_read_per_line = config.max_bytes_read_per_line
+
+    def current_progress(self):
+        """Returns the current progress dictionary."""
+        return self.progress
+
+    def tail(self, sleep_time_ms):
+        """This method incrementally generates lines from a file by waiting for new content from a file.
+        It behaves like the 'tail -f' shell command.
+        
+        Parameters
+        ----------
+        sleep_time_ms: int
+            The unit of time in ms to repetitively sleep when the file has not been created or no new
+            content is available in the file being tailed.
+        
+        Returns
+        -------
+        an incrementally generated list of lines in the file being tailed.
+        """
+        try:
+            sleep_param = sleep_time_ms / 1000
+
+            while not os.path.isfile(self.target_file) and not self.completed_signal.isSet():
+                logging.debug('{} has not yet been created, sleeping {} ms'.format(self.target_file, sleep_time_ms))
+                time.sleep(sleep_param)
+
+            if self.completed_signal.isSet():
+                logging.info('No output has been read to parse progress messages')
+                return
+
+            logging.info('{} has been created, reading contents'.format(self.target_file))
+            target_file_obj = open(self.target_file, 'r')
+            while not self.completed_signal.isSet():
```

Comment:
I don't see how the change you made addresses the original concern.  I'm concerned the `task_completed_signal` will be set immediately upon the completion the subprocess, not allowing the `ProgressWatcher` to report progress updates that are remaining.